src/django_components/component_registry.py

```diff
@@ -14,49 +16,274 @@ class NotRegistered(Exception):
-        self._registry: Dict[str, Type["component.Component"]] = {}  # component name -> component_class mapping
+    """
+    Manages which components can be used in the template tags.
+
+    Each ComponentRegistry instance is associated with an instance
+    of Django's Library. So when you register or unregister a component
+    to/from a component registry, behind the scenes the registry
+    automatically adds/removes the component's template tag to/from
+    the Library.
+
+    The Library instance can be set at instantiation. If omitted, then
+    the default Library instance from django_components is used. The
+    Library instance can be accessed under `library` attribute.
+
+    Example:
+
+    ```py
+    # Use with default Library
+    registry = ComponentRegistry()
+
+    # Or a custom one
+    my_lib = Library()
+    registry = ComponentRegistry(library=my_lib)
+
+    # Usage
+    registry.register("button", ButtonComponent)
+    registry.register("card", CardComponent)
+    registry.all()
+    registry.clear()
+    registry.get()
+    ```
+    """
+
+    def __init__(self, library: Optional[Library] = None) -> None:
+        self._registry: Dict[str, ComponentRegistryEntry] = {}  # component name -> component_entry mapping
+        self._tags: Dict[str, Set[str]] = {}  # tag -> list[component names]
+        self._library = library
 
-    def register(self, name: str, component: Type["component.Component"]) -> None:
+    @property
+    def library(self) -> Library:
+        """
+        The template tag library with which the component registry is associated.
+        """
+        # Lazily use the default library if none was passed
+        if self._library is not None:
+            lib = self._library
+        else:
+            from django_components.templatetags.component_tags import register as tag_library
+
+            lib = self._library = tag_library
+        return lib
+
+    def register(self, name: str, component: Type["Component"]) -> None:
+        """
+        Register a component with this registry under the given name.
+
+        A component MUST be registered before it can be used in a template such as:
+        ```django
+        {% component "my_comp" %}{% endcomponent %}
+        ```
+
+        Raises `AlreadyRegistered` if a different component was already registered
+        under the same name.
+
+        Example:
+
+        ```py
+        registry.register("button", ButtonComponent)
+        ```
+        """
         existing_component = self._registry.get(name)
-        if existing_component and existing_component._class_hash != component._class_hash:
+        if existing_component and existing_component.cls._class_hash != component._class_hash:
             raise AlreadyRegistered('The component "%s" has already been registered' % name)
-        self._registry[name] = component
+
+        used_tags = ["component"]
+
+        # Keep track of which components use which tags, because multiple components may
```

Comment:
What this part does is that when a user defines [custom component tags](https://github.com/EmilStenstrom/django-components/issues/527) , e.g. `my_comp` and `#my_comp`, then we'd add these template tags to the underlying `django.template.Library`.

This way, so it's suficient for users to load only our template tag library (e.g. with `{% load component_tags %}`), and they can still use their custom tags.

And then, bit lower down (https://github.com/EmilStenstrom/django-components/pull/566/files#diff-a5dcf86b1b1bd690092e8ed60318835a1ce9ed2c62befcc05d43dfd4459393b6R150), we do the reverse, and remove any registered component-specific template tags, when the last component using said tag(s) is unregistered.